icicle_v3/backend/cpu/include/ntt_template.h

```diff
@@ -0,0 +1,264 @@
+#include "icicle/ntt.h"
+#include "icicle/errors.h"
+#include "icicle/runtime.h"
+#include "icicle/utils/log.h"
+
+#include "icicle/fields/field_config.h"
+#include "icicle/utils/log.h"
+
+#include <vector>
+#include <algorithm>
+#include <iostream>
+#include <cmath>
+#include <cstdint>
+
+using namespace field_config;
+using namespace icicle;
+namespace ntt_template {
+  template <typename S>
+  class CpuNttDomain
+  {
+    // TODO - coset, mixed-radix NTT
+    int max_size = 0;
+    int max_log_size = 0;
+    S* twiddles = nullptr;
+
+  public:
+    template <typename U>
+    friend eIcicleError
+    cpu_ntt_init_domain(const Device& device, const U& primitive_root, const NTTInitDomainConfig& config);
+
+    template <typename U>
+    friend eIcicleError generate_twiddles(const U& primitive_root, U* twiddles, int logn);
+
+    template <typename U>
+    friend eIcicleError cpu_ntt_release_domain(const Device& device);
+
+    template <typename U, typename E>
+    friend eIcicleError
+    cpu_ntt_ref(const Device& device, const E* input, int size, NTTDir dir, NTTConfig<S>& config, E* output);
+
+    template <typename U, typename E>
+    friend eIcicleError
+    cpu_ntt(const Device& device, const E* input, int size, NTTDir dir, NTTConfig<S>& config, E* output);
+
+    S* get_twiddles() { return twiddles; }
```

Comment:
make it 'const S*' since it should be read only right?

Reply:
Right, fixed